lib/orderbook/MatchingEngine.ts

```diff
@@ -66,34 +101,140 @@ class MatchingEngine {
     };
   }
 
+  /**
+   * Add peer order
+   * @returns false if it's a duplicated order, otherwise true
+   */
+  public addPeerOrder = (order: StampedPeerOrder): boolean => {
+    return this.addOrder(order, this.peerOrders);
+  }
+
   /**
    * Match an order against its opposite queue, and optionally add the unmatched portion of the order to the queue.
    * @param discardRemaining whether to discard any unmatched portion of the order rather than add it to the queue
    * @returns a [[MatchingResult]] with the matches as well as the remaining, unmatched portion of the order
    */
   public matchOrAddOwnOrder = (order: StampedOwnOrder, discardRemaining: boolean): matchingEngine.MatchingResult => {
-    const isBuyOrder = order.quantity > 0;
-    const addTo = isBuyOrder ? this.buyOrders : this.sellOrders;
-
     const matchingResult = this.match(order);
     if (matchingResult.remainingOrder && !discardRemaining) {
-      addTo.add(matchingResult.remainingOrder);
+      this.addOwnOrder(matchingResult.remainingOrder);
     }
 
     return matchingResult;
   }
 
+  private addOwnOrder = (order: StampedOwnOrder): boolean => {
+    return this.addOrder(order, this.ownOrders);
+  }
+
+  private addOrder = (order: StampedOrder, lists: OrderSidesLists<StampedOrder>): boolean => {
+    const isBuyOrder = order.quantity > 0;
+    const list = isBuyOrder ? lists.buy : lists.sell;
+    const queue = isBuyOrder ? this.queues.buy : this.queues.sell;
+
+    if (list.has(order.id)) {
+      return false;
+    } else {
+      list.set(order.id, order);
+      queue.add(order);
+      return true;
+    }
+  }
+
+  public removePeerOrderQuantity = (orderId: string, quantityToDecrease?: number): StampedPeerOrder | undefined => {
+    const order = this.peerOrders.buy.get(orderId) || this.peerOrders.sell.get(orderId);
+    if (!order) {
+      return;
+    }
+
+    if (quantityToDecrease && quantityToDecrease < Math.abs(order.quantity)) {
+      order.quantity = order.quantity - quantityToDecrease;
+      return { ...order, quantity: quantityToDecrease };
```

Comment:
Why are you returning `quantityToDecrease`? It is a parameter so the function calling `removePeerOrderQuantity` will know it anyways. Returning what is left is maybe more sensible.

Reply:
This is how it was before my changes. It's basically simulating the order you remove, and it's being used for the peerOrder invalidation (`subscribePeerOrders`).

Reply:
I agree that it's a weird pattern though to return a parameter, maybe add some comments here explaining what's going on. Which (if I understand correctly), is that it's reducing the quantity of the peer order when the peer order is not fully filled, and returning a copy of it with the quantity that was filled.